Makefile

```
@@ -1,41 +1,182 @@
 
-.PHONY: default
-default: build tests
+VARIABLE+=DEV
+HELP_DEV=which device to flash/monitor
+-include .dev
+DEV?=none
+
+.DEFAULT_GOAL=default
 
 .PHONY: help
 help:
-	@echo "--- standard targets ---"
-	@echo "make help      | shows this help"
-	@echo "make default   | starts build and tests"
-	@echo "make setup     | does nothing"
-	@echo "make all       | starts all builds and tests"
-	@echo "make clean     | cleans build files"
-	@echo "make distclean | removes everything generated"
-	@echo
-	@echo "--- specific targets ---"
-	@echo "make build     | just builds "
-	@echo "make tests     | just tests "
+	@echo "--- help ---"
+	@[ -n "$(HELP_$(.DEFAULT_GOAL))" ] && echo -e "make\n\t$(HELP_$(.DEFAULT_GOAL))" || true;
+	@$(foreach \
+		target, \
+		$(TARGET), \
+		echo "make $(target)"; \
+		[ -n "$(HELP_$(target))" ] && echo -e "\t$(HELP_$(target))"; \
+	)
+	@$(foreach \
+		target, \
+		$(TARGET), \
+		[ -n "$(TARGET_$(target))" ] && echo -e "\n--- $(target) sub targets ---"  || true; \
+		$(foreach \
+			subtarget, \
+			$(TARGET_$(target)), \
+				echo "make $(subtarget)"; \
+				[ -n "$(HELP_$(subtarget))" ] && echo -e "\t$(HELP_$(subtarget))" || true; \
+		) \
+	)
+	@[ -n "$(VARIABLE)" ] && echo -e "\n--- VARIABLES ---"  || true;
+	@$(foreach \
+		variable, \
+		$(VARIABLE), \
+		echo "$(variable)"; \
+		[ -n "$(HELP_$(variable))" ] && echo -e "\t$(HELP_$(variable))" || true; \
+		echo -e "\t(currently: $($(variable)))"; \
+	)
+
+
+TARGET += default
+HELP_default  = builds project
+.PHONY: default
+default: build
 
+TARGET += all
+HELP_all = builds, flashes and starts monitor
 .PHONY: all
-all:
-	@true
+all: build flash monitor
 
-.PHONY: setup
-setup:
-	@true
+TARGET += check
+HELP_check = checks if everything is correctly setup
+.PHONY: check
+check: check-build check-flash check-monitor
 
+TARGET += clean
+HELP_clean = cleans up project
 .PHONY: clean
-clean:
-	@true
+clean: clean-build clean-flash clean-monitor clean-dev
+	@make --no-print-directory -C docker clean
 
+TARGET += distclean
+HELP_distclean = removes all generated files
 .PHONY: distclean
-distclean:
-	@true
+distclean: clean distclean-build distclean-flash distclean-monitor distclean-dev
+	@make --no-print-directory -C docker distclean
+
+### build targets ###
 
+TARGET += build
+HELP_build = builds project
 .PHONY: build
-build:
+build: | check-build
+	@make --no-print-directory -C docker pio EXEC="pio run"
+
+.PHONY: check-build
+check-build: check-docker
+
+.PHONY: clean-build
+clean-build:
+	@make --no-print-directory -C docker pio EXEC="pio run -t clean"
+
+.PHONY: distclean-build
+distclean-build: clean-build
+	rm -rf .pio
+
+### flash targets ###
+
+TARGET += flash
+HELP_flash = flashes project to esp
```

please add info on how to pass dev optionally

---

target has already `check-dev` as dep, which includes instructions

---

But it does not show any infos on dev if ```make help``` is called

---

```--- VARIABLES ---
DEV
	which device to flash/monitor
	(currently: none)
```
?

---

That's my point. You have to fail to get information on how to use the make command.

why not provide some information like 
```suggestion
HELP_flash = flashes project to esp. Use DEV=path to provide  path to device or use make dev
```

---

now i get it.